[SPARK-25108][SQL] Fix the show method to display the wide character alignment problem

xuejianbest opened the pull request:
This is not a perfect solution. It is designed to minimize complexity on the basis of solving problems.

It is effective for English, Chinese characters, Japanese, Korean and so on.

```scala
before:
+---+---------------------------+-------------+
|id |中国                         |s2           |
+---+---------------------------+-------------+
|1  |ab                         |[a]          |
|2  |null                       |[中国, abc]    |
|3  |ab1                        |[hello world]|
|4  |か行 きゃ(kya) きゅ(kyu) きょ(kyo) |[“中国]        |
|5  |中国（你好）a                    |[“中（国）, 312] |
|6  |中国山(东)服务区                  |[“中(国）]      |
|7  |中国山东服务区                    |[中(国)]       |
|8  |                           |[中国]         |
+---+---------------------------+-------------+

after:
+---+-----------------------------------+----------------+
|id |中国                               |s2              |
+---+-----------------------------------+----------------+
|1  |ab                                 |[a]             |
|2  |null                               |[中国, abc]     |
|3  |ab1                                |[hello world]   |
|4  |か行 きゃ(kya) きゅ(kyu) きょ(kyo) |[“中国]         |
|5  |中国（你好）a                      |[“中（国）, 312]|
|6  |中国山(东)服务区                   |[“中(国）]      |
|7  |中国山东服务区                     |[中(国)]        |
|8  |                                   |[中国]          |
+---+-----------------------------------+----------------+
```

## What changes were proposed in this pull request?

When there are wide characters such as Chinese characters or Japanese characters in the data, the show method has a alignment problem. 
Try to fix this problem.

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

![image](https://user-images.githubusercontent.com/13044869/44250564-69f6b400-a227-11e8-88b2-6cf6960377ff.png)

Please review http://spark.apache.org/contributing.html before opening a pull request.

xuejianbest committed:
Fix the show method to display the wide character alignment problem

before:
+---+---------------------------+-------------+
|id |中国                         |s2           |
+---+---------------------------+-------------+
|1  |ab                         |[a]          |
|2  |null                       |[中国, abc]    |
|3  |ab1                        |[hello world]|
|4  |か行 きゃ(kya) きゅ(kyu) きょ(kyo) |[“中国]        |
|5  |中国（你好）a                    |[“中（国）, 312] |
|6  |中国山(东)服务区                  |[“中(国）]      |
|7  |中国山东服务区                    |[中(国)]       |
|8  |                           |[中国]         |
+---+---------------------------+-------------+

after:
+---+--------------------------+----------------+
| id|                      中国|              s2|
+---+--------------------------+----------------+
|  1|                        ab|             [a]|
|  2|                      null|     [中国, abc]|
|  3|                       ab1|   [hello world]|
|  4|か行 きゃ(kya) きゅ(kyu...|         [“中国]|
|  5|             中国（你好）a|[“中（国）, 312]|
|  6|          中国山(东)服务区|      [“中(国）]|
|  7|            中国山东服务区|        [中(国)]|
|  8|                          |          [中国]|
+---+--------------------------+----------------+

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

xuejianbest committed:
Merge branch 'master' into master

kiszk commented:
Since this change does not look minor, could you please create a JIRA entry?

kiszk commented:
I created [a PR](https://github.com/apache/spark/pull/16086) to solve the same problem.
Can this PR handle [East Asian Width](http://www.unicode.org/Public/UCD/latest/ucd/EastAsianWidth.txt) correctly?

xuejianbest commented:
After testing, it is found that regular expressions are changed to the following.
`val regex = """[^\x00-\u2e39]""".r`

xuejianbest committed:
Modifying regular expressions for matching narrow characters

kiszk commented:
@xuejianbest Could you please create a JIRA entry and add test cases to this PR?

xuejianbest commented:
Below is the display effect I tested on Xshell5.
[https://issues.apache.org/jira/secure/attachment/12935462/show.bmp](url)

xuejianbest commented:
![df.show](https://issues.apache.org/jira/secure/attachment/12935462/show.bmp)

kiszk commented:
Thank you for creating a JIRA entry and for putting the result. The test case is not available yet.

srowen commented on sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:
```diff
@@ -294,23 +294,24 @@ class Dataset[T] private[sql](
     // We set a minimum column width at '3'
     val minimumColWidth = 3
 
+    val regex = """[^\x00-\u2e39]""".r
```
This could use a comment and slightly better name for the variable? I also wonder if a regex is a little bit slow for scanning every character.

However it's not clear this definition is accurate enough. According to things like https://stackoverflow.com/questions/13505075/analyzing-full-width-or-half-width-character-in-java we're really looking for the concept of "fullwidth" East Asian characters. The answer there provides a somewhat more precise definition, though others say you need something like icu4j for a proper definition. 

Maybe at least adopt the alternative proposed in the SO answer? It's not necessary to be perfect here, as it's a cosmetic issue.

kiszk replied:
I agree with @srowen's comment. To make it clear, could you please add tests for more characters?

srowen commented:
Ping @xuejianbest

xuejianbest committed:
Fix the show method to display full width characters.

Modify regular expressions to make them more precise.
Modify variable names and add comment.

xuejianbest commented:
I looked at all the 0x0000-0xFFFF characters (unicode) and showed them under Xshell, then found all the full width characters. Get the latest regular expression.
A new version has been submitted, and the variable name has been changed and comments added, please see below.
@srowen @kiszk 

A new version is submitted, and the variable name is changed and an annotation is added. See below.

kiszk commented:
@xuejianbest thank you for your update. Would it be possible to commit test cases, too?

srowen commented on sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:
```diff
@@ -294,23 +294,25 @@ class Dataset[T] private[sql](
     // We set a minimum column width at '3'
     val minimumColWidth = 3
 
+    //Regular expression matching full width characters
```
Very small nit that might fail scalastyle -- space after comment slashes

xuejianbest replied:
OK, thanks.

srowen commented on sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:
```diff
@@ -294,23 +294,25 @@ class Dataset[T] private[sql](
     val minimumColWidth = 3
 
+    //Regular expression matching full width characters
+    val fullWidthRegex = """[\u1100-\u115F\u2E80-\uA4CF\uAC00-\uD7A3\uF900-\uFAFF\uFE10-\uFE19\uFE30-\uFE6F\uFF00-\uFF60\uFFE0-\uFFE6]""".r
```
I wonder if detecting this using a regex for each character is slow? probably not meaningfully enough to care about, but I wonder.

xuejianbest replied:
I generated 1000 strings, each consisting of 1000 characters with a random unicode of 0x0000-0xFFFF. (a total of 1 million characters.)
Then use this regular expression to find the full width character of these strings. 
I tested 100 rounds and then averaged.
It takes 49 milliseconds to complete matching all 1000 strings.

srowen commented on sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:
```diff
@@ -294,23 +294,25 @@ class Dataset[T] private[sql](
       for (row <- rows) {
         for ((cell, i) <- row.zipWithIndex) {
-          colWidths(i) = math.max(colWidths(i), cell.length)
+          colWidths(i) = math.max(colWidths(i), cell.length + fullWidthRegex.findAllIn(cell).size)
```
What about a utility method for the `x.length + fullWidthRegex.findAllIn(x).size` expression that recurs here?

xuejianbest replied:
I committed a new version. See if this is appropriate please ?
@srowen

xuejianbest committed:
Write a utility method to get the number of half width of string

Putting `fullWidthRegex` outside `stringHalfWidth` is to reduce the cost of compiling regular expressions.
Since compilation is expensive, frequently used Regexes should be constructed once.

maropu commented on sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:
```diff
@@ -294,23 +294,29 @@ class Dataset[T] private[sql](
+    // The number of half width of a string
+    def stringHalfWidth = (str: String) => {
+      str.length + fullWidthRegex.findAllIn(str).size
+    }
```
better to move this method into `util.Utils` or something as a helper function?

maropu replied:
better to add tests for the helper function, too.

xuejianbest replied:
Moved it into `util.Utils`.

maropu commented on sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:
```diff
@@ -294,23 +294,29 @@ class Dataset[T] private[sql](
     val minimumColWidth = 3
 
+    // Regular expression matching full width characters
+    val fullWidthRegex = """[\u1100-\u115F\u2E80-\uA4CF\uAC00-\uD7A3\uF900-\uFAFF\uFE10-\uFE19\uFE30-\uFE6F\uFF00-\uFF60\uFFE0-\uFFE6]""".r
```
This line goes over the limit, 100.

xuejianbest committed:
Move the stringHalfWidth method into util.Utils

Move the stringHalfWidth method into util.Utils and add tests for it.

xuejianbest commented:
I have moved the method of calculating the half width of a string into util.Utils object.
@maropu @srowen

srowen commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -2794,6 +2794,27 @@ private[spark] object Utils extends Logging {
+   * A full width character occupies two half widths
+   */
+  def stringHalfWidth(str: String): Int = {
+    if(str == null) 0 else str.length + fullWidthRegex.findAllIn(str).size
```
Nit: space before if

srowen commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -2794,6 +2794,27 @@ private[spark] object Utils extends Logging {
+  /**
+   * Regular expression matching full width characters
+   */
+  private lazy val fullWidthRegex = ("""[""" +
```
Doesn't need to be lazy

srowen commented on core/src/test/scala/org/apache/spark/util/UtilsSuite.scala:
```diff
@@ -1184,6 +1184,25 @@ class UtilsSuite extends SparkFunSuite with ResetSystemProperties with Logging {
+    assert(Utils.stringHalfWidth("ab c") == 4)
+    assert(Utils.stringHalfWidth("1098") == 4)
+    assert(Utils.stringHalfWidth("mø") == 2)
+    assert(Utils.stringHalfWidth("γύρ") == 3)
```
I think it is OK to type Unicode chars into source files because git and Scala ought to be set up to encode and read the source as UTF8. I have avoided it in the past and used Unicode escapes (but put actual string in comment) to be safe. I don't feel strongly. I think we have other instances of Unicode in the source. Just a comment.

xuejianbest replied:
Do you mean that the string here needs to be converted to a unicode escape form?

srowen replied:
Don't change it. There are Unicode characters elsewhere in the source, and it's clearer this way.

xuejianbest replied:
I found similar code in the `org.apache.commons.lang3.StringUtils` class.
So the same test was added.
In StringUtils 6934 line is comment:
`* StringUtils.isNumeric("\u0967\u0968\u0969")  = true`

srowen commented on sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala:
```diff
@@ -301,16 +301,16 @@ class Dataset[T] private[sql](
       val paddedRows = rows.map { row =>
         row.zipWithIndex.map { case (cell, i) =>
           if (truncate > 0) {
-            StringUtils.leftPad(cell, colWidths(i))
```
Oh why not use this method?

xuejianbest replied:
This method fills the string to the specified length.
But what you need to do here is to increase the string by the specified length by padding.
If you want to use this function, code programming:
`StringUtils.leftPad(cell, colWidths(i) - Utils.stringHalfWidth(cell) + cell.length)`
I think it is more difficult to understand.

srowen replied:
OK that's fine.

maropu commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -2794,6 +2794,27 @@ private[spark] object Utils extends Logging {
+    """\uFE30-\uFE6F""" +
+    """\uFF00-\uFF60""" +
+    """\uFFE0-\uFFE6""" +
+    """]""").r
```
super nit: add blank line

maropu commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -2794,6 +2794,27 @@ private[spark] object Utils extends Logging {
+  /**
+   * Return the number of half width of a string
+   * A full width character occupies two half widths
+   */
```
How about this?
```
  /**
   * Return the number of half widths in a given string. Note that a full width character
   * occupies two half widths.
   */
```

xuejianbest replied:
OK, thinks

xuejianbest committed:
Formatted code and non-functional changes

xuejianbest committed:
Add tests in DatasetSuite

xuejianbest committed:
Modified variable name in DatasetSuite

SparkQA commented:
**[Test build #4305 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4305/testReport)** for PR 22048 at commit [`3649de5`](https://github.com/apache/spark/commit/3649de50235cd19cfea2c3d88d1ccfb18ea8893a).

SparkQA commented:
**[Test build #4305 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4305/testReport)** for PR 22048 at commit [`3649de5`](https://github.com/apache/spark/commit/3649de50235cd19cfea2c3d88d1ccfb18ea8893a).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

xuejianbest commented:
Why did it fail?
What else should I do?

xuejianbest commented:
Please help me @srowen 
I don't know where I need to make changes.

srowen commented:
Have a look at the test results, which show the style problem it flags:
https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4305/console

```
[error] /home/jenkins/workspace/NewSparkPullRequestBuilder@2/core/src/main/scala/org/apache/spark/util/Utils.scala:2803:4: nonascii.message
[error] /home/jenkins/workspace/NewSparkPullRequestBuilder@2/core/src/main/scala/org/apache/spark/util/Utils.scala:2804:4: nonascii.message
[error] /home/jenkins/workspace/NewSparkPullRequestBuilder@2/core/src/main/scala/org/apache/spark/util/Utils.scala:2805:4: nonascii.message
[error] /home/jenkins/workspace/NewSparkPullRequestBuilder@2/core/src/main/scala/org/apache/spark/util/Utils.scala:2806:4: nonascii.message
[error] /home/jenkins/workspace/NewSparkPullRequestBuilder@2/core/src/main/scala/org/apache/spark/util/Utils.scala:2807:4: nonascii.message
[error] /home/jenkins/workspace/NewSparkPullRequestBuilder@2/core/src/main/scala/org/apache/spark/util/Utils.scala:2808:4: nonascii.message
[error] /home/jenkins/workspace/NewSparkPullRequestBuilder@2/core/src/main/scala/org/apache/spark/util/Utils.scala:2809:4: nonascii.message
[error] /home/jenkins/workspace/NewSparkPullRequestBuilder@2/core/src/main/scala/org/apache/spark/util/Utils.scala:2810:4: nonascii.message
```

So it doesn't like non-ASCII chars after all! I wonder how they exist in the other parts of the source. I checked and other parts just disable the check for sections of code where it's appropriate to use unicode chars.

Just wrap these parts of the code in:

```
// scalastyle:off nonascii
...
// scalastyle:on nonascii
```

Sorry, didn't know this was needed.

maropu commented:
style errors, we cannot include non-ascii characters in files;
```
Scalastyle checks failed at following occurrences:
[error] /home/jenkins/workspace/NewSparkPullRequestBuilder@2/core/src/main/scala/org/apache/spark/util/Utils.scala:2803:4: nonascii.message
[error] /home/jenkins/workspace/NewSparkPullRequestBuilder@2/core/src/main/scala/org/apache/spark/util/Utils.scala:2804:4: nonascii.message
[error] /home/jenkins/workspace/NewSparkPullRequestBuilder@2/core/src/main/scala/org/apache/spark/util/Utils.scala:2805:4: nonascii.message
[error] /home/jenkins/workspace/NewSparkPullRequestBuilder@2/core/src/main/scala/org/apache/spark/util/Utils.scala:2806:4: nonascii.message
[error] /home/jenkins/workspace/NewSparkPullRequestBuilder@2/core/src/main/scala/org/apache/spark/util/Utils.scala:2807:4: nonascii.message
[error] /home/jenkins/workspace/NewSparkPullRequestBuilder@2/core/src/main/scala/org/apache/spark/util/Utils.scala:2808:4: nonascii.message
[error] /home/jenkins/workspace/NewSparkPullRequestBuilder@2/core/src/main/scala/org/apache/spark/util/Utils.scala:2809:4: nonascii.message
[error] /home/jenkins/workspace/NewSparkPullRequestBuilder@2/core/src/main/scala/org/apache/spark/util/Utils.scala:2810:4: nonascii.message
[error] Total time: 16 s, completed Aug 30, 2018 6:48:19 AM
[error] running /home/jenkins/workspace/NewSparkPullRequestBuilder@2/dev/lint-scala ; received return code 1
Attempting to post to Github...
 > Post successful.
```

maropu commented:
+1 for adding `scalastyle:off nonascii` in that part.

xuejianbest committed:
Adding scalastyle:off nonascii in code

To solve style errors because of include non-ascii characters in files.
Disable the check for sections of code where it's appropriate to use unicode chars.

xuejianbest commented:
I see. A new commit has been done.
Thinks.

kiszk commented:
retest this please

srowen commented:
Hm! so the warning triggers on unicode escape sequences? That seems like a bug in the style checker. Oh well, yes keep them. I'll retest ...

SparkQA commented:
**[Test build #4329 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4329/testReport)** for PR 22048 at commit [`45ac272`](https://github.com/apache/spark/commit/45ac272ca667f3330f7b550b463b23c284d9eadf).

SparkQA commented:
**[Test build #4329 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4329/testReport)** for PR 22048 at commit [`45ac272`](https://github.com/apache/spark/commit/45ac272ca667f3330f7b550b463b23c284d9eadf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

kiszk commented:
LGTM, then let me ask @gatorsmile

gatorsmile commented on core/src/main/scala/org/apache/spark/util/Utils.scala:
```diff
@@ -2794,6 +2794,30 @@ private[spark] object Utils extends Logging {
+    """\uFE10-\uFE19""" +
+    """\uFE30-\uFE6F""" +
+    """\uFF00-\uFF60""" +
+    """\uFFE0-\uFFE6""" +
```
A general question. 
- How to get this Regex list? Any reference?  It sounds like this should be a general problem
- What is the performance impact?

Can you answer them and post them in the PR description?

xuejianbest replied:
- How to get this Regex list? Any reference? It sounds like this should be a general problem

I looked at all the 0x0000-0xFFFF characters (unicode) and showed them under Xshell, then found all the full width characters. Get the regular expression.


- What is the performance impact?

I generated 1000 strings, each consisting of 1000 characters with a random unicode of 0x0000-0xFFFF. (a total of 1 million characters.)
Then use this regular expression to find the full width character of these strings.
I tested 100 rounds and then averaged.
It takes 49 milliseconds to complete matching all 1000 strings.

@gatorsmile

xuejianbest replied:
Do I need to merge the above commited into one commit, 
Or add another new commit?
Or change the last commit comments ?
@srowen

srowen commented:
Just add a new commit with a little info about what this does and how you generated it. The commits are squashed anyway.

xuejianbest committed:
Fix the show method to display the full width character alignment problem

Some characters (called full-width characters) occupy two
ordinary characters (called half-width characters) width
when displayed in a terminal such as Xshell.
Therefore, when the Dataset.show() method is called,
if there are full-width characters and half-width characters
at the same time, alignment problems will occur.

This commit adds a method to calculate the number of half
widths in a given string, so that it can get the correct fill
when calling show method, which gives the display a more
perfect alignment.

Performance impact:
Tested a Dataset consisting of 100 rows, each row has two columns,
one column is the index (0-99), and the other column is a random
string of length 100 characters, and then the showString display
is called separately.
The original show method (w/o this patch) took about 42ms, and the
improved time took about 46ms, the performance was about 10% worse.

xuejianbest commented:
I see, Thinks.

srowen commented:
Merging to master and 2.4, though for some reason I can't back-port with the merge script right now. Will try later.

dongjoon-hyun commented:
@srowen . Sure. I'll try to backport this.

dongjoon-hyun commented:
Ur, it seems that I'm facing the same situation. In this case, should we just cherry-pick to `branch-2.4`?

srowen commented:
OK, it isn't just me. We'll see if it's a broader problem.

I manually cherry-picked to branch-2.4
